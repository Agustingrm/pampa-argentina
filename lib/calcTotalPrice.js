export default function calcTotalPrice(cart) {
  return cart.reduce((acc, cartItem) => {
    if (!cartItem.product) return acc;
    return acc + cartItem.quantity * cartItem.product.price;
  }, 0);
}
