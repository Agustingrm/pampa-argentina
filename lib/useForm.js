import { useEffect, useState } from "react";

export default function useForm(initial = {}) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);
  const initialValues = Object.values(initial).join("");

  useEffect(() => {
    // This is used in update, when the query finished loading, you need to put the values in the inputs
    setInputs(initial);
  }, [initialValues]);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === "number") {
      value = parseInt(value);
    }
    if (type === "file") {
      value = e.target.files;
    }

    setInputs({
      // Copy the existing state
      ...inputs,
      // Brackets here is a dynamic value
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key, value]) => [key, ""]));
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
