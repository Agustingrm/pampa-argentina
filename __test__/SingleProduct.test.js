import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import SingleProduct, { SINGLE_ITEM_QUERY } from "../components/SingleProduct";
import { fakeItem } from "../lib/testUtils";

const product = fakeItem();

const mocks = [
  {
    // When someone requests this query and variable
    request: {
      query: SINGLE_ITEM_QUERY,
      variables: {
        id: "123",
      },
    },
    // Return this data
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe("<SingleProduct />", () => {
  it("renders with proper data", async () => {
    // We need to make some fake data
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    // Wait for the test ID to show up
    await screen.findAllByTestId("singleProduct");
    expect(container).toMatchSnapshot();
  });

  it("Errors out when an item is not found", async () => {
    const errorMock = [
      {
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: {
            id: "123",
          },
        },
        result: {
          errors: [{ message: "Item not found" }],
        },
      },
    ];
    const { container } = render(
      <MockedProvider mocks={errorMock}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    await screen.findAllByTestId("graphql-error");
    expect(container).toHaveTextContent("Error");
    expect(container).toHaveTextContent("Item not found");
  });
});
