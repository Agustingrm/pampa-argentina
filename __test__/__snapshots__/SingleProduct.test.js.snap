// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1q67if-0 eAHSXo"
    data-testid="singleProduct"
  >
    <div
      class="details"
    >
      <div>
        <img
          alt="dogs are best"
          src="dog.jpg"
        />
      </div>
      <div>
        <h2>
          dogs are best
        </h2>
        <p>
          dogs
        </p>
      </div>
    </div>
  </div>
</div>
`;
